### PR TITLE
perf(ce): hardware exp2 (ex2.approx) for softmax

### PR DESCRIPTION
**Pull request by justinhh4**
## PR1: CE `exp2` — B200 + H100 benchmarks

Optimization measured against the **frozen original baseline** (`v0_base`, == upstream HEAD). Correctness: full CE suite (177 cases, bf16+fp32, weight/softcap/smoothing/z_loss/ignore_index combos) passes.

### What & why
Replace `tl.exp` with the hardware `ex2.approx` instruction via `e^x = 2^(x·log₂e)`, in both the online-softmax accumulation and the gradient pass.
- **Why:** NVIDIA GPUs have no native `e^x` — `tl.exp` lowers to a multi-instruction accurate sequence, while `tl.exp2` emits a single `ex2.approx.f32`. This is the most frequent FP op in the kernel (once per vocab element × 2 passes), so it cuts total issued instructions.
- **Accuracy:** `ex2.approx` has ~2.4e-7 relative error — within the suite's fp32 `rtol=1e-6` and bf16 `rtol=5e-2`, and softmax normalization further cancels the (correlated) error. Helps every dtype.

### Headline shape (BT=8192, V=128256, full fwd+bwd)

| GPU | dtype | baseline | this PR | latency | throughput |
|---|---|--:|--:|--:|--:|
| H100 | bf16 | 2.583 ms | 2.225 ms | **-13.9%** | +16.1% |
| H100 | fp32 | 4.489 ms | 4.475 ms | **-0.3%** | +0.3% |
| B200 | bf16 | 1.847 ms | 1.687 ms | **-8.6%** | +9.5% |
| B200 | fp32 | 2.085 ms | 1.887 ms | **-9.5%** | +10.5% |

### Across context length (V=128256, BT 1024–65536)

| GPU | dtype | min | avg | max |
|---|---|--:|--:|--:|
| H100 | bf16 | -10% | -13% | -15% |
| H100 | fp32 | -0% | -0% | -1% |
| B200 | bf16 | -4% | -8% | -10% |
| B200 | fp32 | -5% | -8% | -10% |

### Figures
_Latency vs context length and vs vocab, baseline vs this PR, on H100 and B200._

pr1_bf16_bt
pr1_bf16_vocab
pr1_fp32_bt
pr1_fp32_vocab

### CODE REVIEW

File: src/liger_kernel/ops/cross_entropy.py
Comment (vaibhavjindal) [inline src/liger_kernel/ops/cross_entropy.py:1]: We should define 1.4426950408889634 as a constant and use it instead of explicitly writing it multiple times.

**Comment by vaibhavjindal** [comment]
Good find! Added a minor comment.

@Tcc0403 @Mecoli1219 do you see any issues with this change?

File: src/liger_kernel/ops/cross_entropy.py
Comment (justinhh4) [reply src/liger_kernel/ops/cross_entropy.py:1]: Changed it to a constant.

**Review by Tcc0403 — APPROVED** [review]
LGTM

**Review by vaibhavjindal — APPROVED** [review]

**Comment by yueyiming2009** [comment]
A late review but should we have a wrapper to improve readability? @vaibhavjindal @Tcc0403 ?

`@triton.jit
def fast_exp(x):
    """
    Hardware-accelerated base-e exponentiation.
    Uses ex2.approx via the identity e^x = 2^(x * log2(e)).
    """
    LOG2_E = 1.4426950408889634
    return tl.exp2(x * LOG2_E)`

**Comment by Tcc0403** [comment]
I prefer plain exp2 call, leaving developers to decide ordering for numerical concerns. E.g. `exp2((a - b) - log2_e)` or `exp2(a * log2_e - b * log2_e)`
